mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+use std::{
+    io::{Cursor, Seek, SeekFrom},
+    str::FromStr,
+    time::{SystemTime, SystemTimeError, UNIX_EPOCH},
+};
+
+use bincode::{
+    deserialize_from,
+    serialize,
+    serialize_into,
+    serialized_size,
+    Error,
+};
+use crc::{crc32, Hasher32};
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+
+use crate::{
+    bdev::nexus::{
+        nexus_child::{ChildError, ChildIoError, NexusChild},
+        nexus_label::{GptEntry, GptGuid, Unsigned},
+        nexus_state::NexusConfig,
+    },
+    core::{DmaBuf, DmaError},
+};
+
+#[derive(Debug, Snafu)]
+pub enum MetaDataError {
+    #[snafu(display("Incorrect MetaData header size"))]
+    HeaderSize {},
+    #[snafu(display("Incorrect MetaData header signature"))]
+    HeaderSignature {},
+    #[snafu(display("Incorrect MetaData header checksum"))]
+    HeaderChecksum {},
+    #[snafu(display("Incorrect MetaData index checksum"))]
+    IndexChecksum {},
+    #[snafu(display("Incorrect MetaData configuration object checksum"))]
+    ObjectChecksum {},
+
+    #[snafu(display("MetaData serialization error"))]
+    SerializeError { source: Error },
+    #[snafu(display("MetaData deserialization error"))]
+    DeserializeError { source: Error },
+
+    #[snafu(display("Failed to allocate index: {}", source))]
+    IndexAlloc { source: DmaError },
+    #[snafu(display("Failed to write index: {}", source))]
+    IndexWrite { source: ChildIoError },
+
+    #[snafu(display("MetaData index is inconsistent"))]
+    IndexInconsistent {},
+    #[snafu(display(
+        "Number of objects {} exceeds index size {}",
+        count,
+        size
+    ))]
+    IndexSizeExceeded { count: u32, size: u32 },
+    #[snafu(display("Not enough space left on MetaData partition"))]
+    PartitionSizeExceeded {},
+
+    #[snafu(display("Failed to allocate configuration object: {}", source))]
+    ObjectAlloc { source: DmaError },
+    #[snafu(display("Failed to read configuration object: {}", source))]
+    ObjectRead { source: ChildIoError },
+    #[snafu(display("Failed to write configuration object: {}", source))]
+    ObjectWrite { source: ChildIoError },
+
+    #[snafu(display("MetaData partition is missing or invalid"))]
+    MissingPartition {},
+
+    #[snafu(display("Error calculating timestamp: {}", source))]
+    TimeStampError { source: SystemTimeError },
+
+    #[snafu(display("{}", error))]
+    NexusChildError { error: String },
+}
+
+#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Copy, Clone)]
+pub struct MetaDataHeader {
+    /// Signature identifying this as a MetaDataHeader object
+    pub signature: [u8; 8],
+    /// Size of this header in bytes
```

Comment:
Looks nice with the comments, doesn't it? :+1: Please also add a /// above the struct itself :)